atc/db/check_lifecycle.go

```diff
@@ -0,0 +1,41 @@
+package db
+
+//go:generate counterfeiter . CheckLifecycle
+
+type CheckLifecycle interface {
+	DeleteCompletedChecks() error
+}
+
+type checkLifecycle struct {
+	conn Conn
+}
+
+func NewCheckLifecycle(conn Conn) CheckLifecycle {
+	return &checkLifecycle{
+		conn: conn,
+	}
+}
+
+func (cl *checkLifecycle) DeleteCompletedChecks() error {
+	_, err := cl.conn.Exec(`
+      WITH deletable_builds AS
+      (
+        SELECT id
+        FROM builds b
+        WHERE completed AND resource_id IS NOT NULL
+        AND EXISTS (SELECT * FROM builds b2 WHERE b.resource_id = b2.resource_id AND b.id < b2.id)
+          UNION ALL
+        SELECT id
+        FROM builds b
+        WHERE completed AND resource_type_id IS NOT NULL
+        AND EXISTS (SELECT * FROM builds b2 WHERE b.resource_type_id = b2.resource_type_id AND b.id < b2.id)
+      ),
+      deleted_builds AS
+      (
+        DELETE FROM builds USING deletable_builds WHERE builds.id = deletable_builds.id
+        RETURNING builds.id
+      )
+      DELETE FROM check_build_events USING deletable_builds WHERE build_id = deletable_builds.id
```

Comment:
Oh were we going to use the `deleted_builds` table rather than the `deletable_builds`?

Reply:
I decided to use `deletable_builds` instead in the hopes that Postgres would realize "oh, you don't actually need the `RETURNING builds.id` part from `deleted_builds` and wouldn't store anything. I'm really just using `deleted_builds` as a hacky way to chain deletes in a single query

Reply:
Ohh could we just remove the returning part of the delete then?

Reply:
oh lol I just assumed that you couldn't use a straight `DELETE` in a CTE and didn't bother testing it, but yeah you're totally right. removed

Reply:
Actually I refactored it again to remove the `deletable_builds` CTE and just use a subquery within `deleted_builds`, and added back the `RETURNING id` clause - seems a bit more straightforward that way